mmpose/models/heads/heatmap_heads/heatmap_head.py

```diff
@@ -304,56 +242,18 @@ def forward(self, feats: Tuple[Tensor]) -> Tensor:
                 test_cfg: ConfigType) -> SampleList:
         """Predict results from features."""
-        heatmaps = self.forward(feats)
-        preds = self.decode(heatmaps, batch_data_samples, test_cfg)
-        return preds
-
-    def decode(self, batch_heatmaps: Tensor, batch_data_samples: OptSampleList,
-               test_cfg: ConfigType) -> SampleList:
-        """Decode keypoints from heatmaps."""
-        if self.decoder is None:
-            raise RuntimeError(
-                f'The decoder has not been set in {self.__class__.__name__}. '
-                'Please set the decoder configs in the init parameters to '
-                'enable head methods `head.predict()` and `head.decode()`')
-
-        if batch_data_samples is None:
-            raise ValueError(
-                '`batch_data_samples` is required to decode keypoitns.')
-
-        # TODO: support decoding with tensor data
-        batch_heatmaps_np = _to_numpy(batch_heatmaps)
-        for heatmaps, data_sample in zip(batch_heatmaps_np,
-                                         batch_data_samples):
-            keypoints, scores = self.decoder.decode(heatmaps)
-
-            # Convert the decoded local keypoints (in input space)
-            # to the image coordinate space
-            # Convert keypoint coordinates from input space to image space
-            if 'gt_instances' in data_sample:
-                bbox_centers = data_sample.gt_instances.bbox_centers
-                bbox_scales = data_sample.get_instances.bbox_scales
-                input_size = data_sample.metainfo.input_size
-                keypoints = keypoints / input_size * bbox_scales + \
-                    bbox_centers - 0.5 * bbox_scales
 
-            else:
-                raise ValueError(
-                    '`gt_instances` is required to convert keypoints from'
-                    ' from the heatmap space to the image space.')
+        batch_heatmaps = self.forward(feats)
+        pred_coords = self.decode(batch_heatmaps, batch_data_samples, test_cfg)
 
-            # Store the keypoint predictions in the data sample
-            if 'pred_instances' not in data_sample:
-                data_sample.pred_instances = InstanceData()
-            data_sample.pred_instances.keypoints = keypoints
-            data_sample.pred_instances.keypoint_scores = scores
+        # Whether to visualize the predicted heatmps
+        vis_heatmaps = test_cfg.get('vis_heatmaps', True)
 
-            # Store the heatmap predictions in the data sample
-            if 'pred_fileds' not in data_sample:
-                data_sample.pred_fields = PixelData()
-            data_sample.pred_fields.heatmaps = heatmaps
+        for heatmaps, data_sample in zip(batch_heatmaps, batch_data_samples):
+            if vis_heatmaps:
+                data_sample.pred_fields.heatmaps = heatmaps
 
-        return batch_data_samples
+        return pred_coords
```

Comment:
`pred_coords` can be confusing because it's actually a list of data samples (the input data samples with `pred_instances` added in place) instead of just coordinates. And maybe 'output_heatmaps' is better than 'vis_heatmaps'.

```python
if test_cfg.get('output_heatmaps', True):
    for heatmaps, data_sample in zip(batch_heatmaps, batch_data_samples):
        if 'pred_fields' not in data_sample:
            data_sample.pred_fields = PixelData()
        data_sample.pred_fields.heatmaps = heatmaps
```